File: kotlin/src/main/kotlin/adventofcode2022/Day04.kt
package adventofcode2022

import adventofcode2022.util.runWithStopwatch
import java.io.File
import kotlin.math.sign

private val input = File("./input/day04.txt").readLines()

private fun part1() {
    val numberOfOverlapping = input.stream().filter { pair ->
        val sections = pair.split(",", "-")
        val firstSectionPair = sections[0].toDouble() - sections[2].toDouble()
        val secondSectionPair = sections[1].toDouble() - sections[3].toDouble()
        firstSectionPair.sign + secondSectionPair.sign == 0.0 ||
                (firstSectionPair.sign == 0.0 || secondSectionPair.sign == 0.0)
    }.toList().size
    println("Total number of overlapping: $numberOfOverlapping" )
}

private fun part2() {
    val numberOfAllOverlapping = input.stream().filter {
        val sections = it.split(",", "-")
        val firstSectionRange = (sections[0].toInt()..sections[1].toInt()).toList()
        val secondSectionRange = (sections[2].toInt()..sections[3].toInt()).toList()
        firstSectionRange.intersect(secondSectionRange).isNotEmpty()
    }.toList().size
    println("Total number of all overlapping: $numberOfAllOverlapping" )
}

fun main() {
    runWithStopwatch { part1() }
    runWithStopwatch { part2() }
}

File: python/Day04.py
from utils.Stopwatch import runWithStopwatch

input_lines = open("../input/day04.txt", "r").readlines()


def part1():
    number_of_overlapping = 0
    for line in input_lines:
        sections = line.strip().split(",")
        first = getRange(sections[0])
        second = getRange(sections[1])
        if first.issubset(second) or second.issubset(first):
            number_of_overlapping += 1
    print("Total number of overlapping: " + str(number_of_overlapping))


def getRange(section):
    ids = section.split("-")
    return set(range(int(ids[0]), int(ids[1]) + 1))


def part2():
    number_of_all_overlapping = 0
    for line in input_lines:
        sections = line.strip().split(",")
        first = getRange(sections[0])
        second = getRange(sections[1])
        if first.intersection(second) or second.intersection(first):
            number_of_all_overlapping += 1
    print("Total number of all overlapping: " + str(number_of_all_overlapping))


if __name__ == '__main__':
    runWithStopwatch(part1)
    runWithStopwatch(part2)
